overlord/devicestate/helpers.go

```diff
@@ -41,3 +49,53 @@ func gadgetDataFromInfo(info *snap.Info, constraints *gadget.ModelConstraints) (
 	}
 	return &gadget.GadgetData{Info: gi, RootDir: info.MountDir()}, nil
 }
+
+// partitionFromLabel returns the node of the block device used by the filesystem
```

Comment:
@bboozzoo do we have code already like these helpers because of gadget stuff or are these really new ?

Reply:
We have some code here: https://github.com/snapcore/snapd/blob/master/gadget/device_linux.go#L43 there's FindDeviceForStructure() and FindDeviceForStructureWithFallback(). It locates the partitions using their name/fs label. There's some unhandled scenarios with the fallback, but that's only relevant for structures that cannot be identified by a name or the fs label, or otherwise have no partition table entry, which AFAIU is not the case here.

Reply:
Thanks @bboozzoo. I think it would be easy to reimplement `partitionFromLabel()` to use `gadget.FindDeviceForStructure()` instead of forking `findfs(8)`, along the lines of:
```
func partitionFromLabel(gadgetDir, label string) (string, error) {
        pv, err := gadget.PositionedVolumeFromGadget(gadgetDir)
        if err != nil {
                return "", err
        }
        var targetStructure *gadget.LaidOutStructure
        for _, ps := range pv.LaidOutStructure {
                vs := ps.VolumeStructure
                if vs.Label == label {
                        targetStructure = &ps
                        break
                }
        }
        if targetStructure == nil {
                return "", fmt.Errorf("cannot find structure with label %q", label)
        }
        return gadget.FindDeviceForStructure(targetStructure)
}
```
It's a slightly more intricate alternative that would re-parse the gadget definition but should work equivalently. The helper to obtain the disk block device node based on the partition would stay the same. What do you think?

Reply:
Yes, I think it looks fine.

Reply:
Updated to use a variation of the code outlined in the previous message.